### PR TITLE
HADOOP-15430. hadoop fs -mkdir -p path-ending-with-slash/ fails with s3guard

### PR DESCRIPTION
**Pull request by steveloughran**
path qualification in s3a fs strips any trailing / ; with tests

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 45 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1311 | trunk passed |
| +1 | compile | 38 | trunk passed |
| +1 | checkstyle | 25 | trunk passed |
| +1 | mvnsite | 41 | trunk passed |
| +1 | shadedclient | 824 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 32 | trunk passed |
| 0 | spotbugs | 63 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 61 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 38 | the patch passed |
| +1 | compile | 29 | the patch passed |
| +1 | javac | 29 | the patch passed |
| -0 | checkstyle | 21 | hadoop-tools/hadoop-aws: The patch generated 4 new + 18 unchanged - 0 fixed = 22 total (was 18) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 878 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | the patch passed |
| +1 | findbugs | 64 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 91 | hadoop-aws in the patch passed. |
| +1 | asflicense | 35 | The patch does not generate ASF License warnings. |
| | | 3707 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1646 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 00309628183a 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ec86f42 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/1/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 52 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1178 | trunk passed |
| +1 | compile | 37 | trunk passed |
| +1 | checkstyle | 28 | trunk passed |
| +1 | mvnsite | 41 | trunk passed |
| +1 | shadedclient | 841 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 30 | trunk passed |
| 0 | spotbugs | 61 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 59 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 34 | the patch passed |
| +1 | compile | 29 | the patch passed |
| +1 | javac | 29 | the patch passed |
| -0 | checkstyle | 21 | hadoop-tools/hadoop-aws: The patch generated 4 new + 18 unchanged - 0 fixed = 22 total (was 18) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 837 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 24 | the patch passed |
| +1 | findbugs | 62 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 87 | hadoop-aws in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 3543 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1646 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 9e6a187f2297 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ec86f42 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/2/testReport/ |
| Max. process+thread count | 417 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 1718 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 4 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1280 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 23 | trunk passed |
| +1 | mvnsite | 36 | trunk passed |
| +1 | shadedclient | 868 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 26 | trunk passed |
| 0 | spotbugs | 59 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 28 | the patch passed |
| +1 | javac | 28 | the patch passed |
| -0 | checkstyle | 18 | hadoop-tools/hadoop-aws: The patch generated 4 new + 18 unchanged - 0 fixed = 22 total (was 18) |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 871 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 23 | the patch passed |
| +1 | findbugs | 60 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 71 | hadoop-aws in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 5295 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1646 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 5ed9b24255fe 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 8c9173c |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/3/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/3/testReport/ |
| Max. process+thread count | 427 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1646/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/PathMetadataDynamoDBTranslation.java [hunk 1]
```diff
@@ -318,8 +318,20 @@ public static String pathToParentKey(Path path) {
...
-    return new PrimaryKey(PARENT, pathToParentKey(path.getParent()), CHILD,
-        path.getName());
+    String childName = path.getName();
+    PrimaryKey key = new PrimaryKey(PARENT,
+        pathToParentKey(path.getParent()), CHILD,
+        childName);
+    for (KeyAttribute attr : key.getComponents()) {
+      String name = attr.getName();
+      Object v = attr.getValue();
+      Preconditions.checkNotNull(v,
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/PathMetadataDynamoDBTranslation.java]: should move to Objects.notNull()

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AMiscOperations.java [hunk 2]
```diff
@@ -241,4 +242,117 @@ public void testS3AToStringUnitialized() throws Throwable {
...
+   * Verify that an object's string value path has a single trailing / symbol;
+   * returns the object.
+   * @param role role for error messages
+   * @param o object
+   * @param <T> type of object
+   * @return the object.
+   */
+  private static <T> T verifyTrailingSlash(String role, T o) {
+    String s = o.toString();
+    assertTrue(role + " lacks trailing slash " + s,
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AMiscOperations.java]: could probably do this with assertJ better

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java [hunk 3]
```diff
@@ -1003,7 +1003,7 @@ private void compareDir(FileStatus msDir, FileStatus s3Dir,
      * @throws IOException on I/O errors.
      */
     private void compareRoot(Path path, PrintStream out) throws IOException {
-      Path qualified = getFilesystem().qualify(path);
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java]: We don't need to change this. qualify() will call makeQualified.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java [hunk 4]
```diff
@@ -1034,7 +1034,7 @@ public int run(String[] args, PrintStream out) throws IOException {
       } else {
         root = new Path(uri.getPath());
       }
-      root = getFilesystem().qualify(root);
+      root = getFilesystem().makeQualified(root);
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java]: We don't need to change this. qualify() will call makeQualified.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AMiscOperations.java (hunk 2 above)
Comment (bgaborg) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AMiscOperations.java]: it's good enough, readable.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/AbstractS3GuardToolTestBase.java [hunk 5]
```diff
@@ -231,7 +231,7 @@ protected void createFile(Path path, boolean onS3, boolean onMetadataStore)
       ContractTestUtils.touch(fs, path);
     } else if (onMetadataStore) {
       S3AFileStatus status = new S3AFileStatus(100L, System.currentTimeMillis(),
-          fs.qualify(path), 512L, "hdfs", null, null);
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/AbstractS3GuardToolTestBase.java]: we do we need to change this? to remember in the future that we changed?

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/PathMetadataDynamoDBTranslation.java (hunk 1 above)
Comment (bgaborg) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/PathMetadataDynamoDBTranslation.java]: maybe, but it's good enough and easily readable.

**Comment by steveloughran** [comment]
I'm using makeQualified(Path path) over qualify(Path) because the first one is in FileSystem and is the public API we should be using, so: won't change those new uses

**Comment by whenamanlies** [comment]
Is it possible that Path.toUri() is missing .getPath() as it seem to have broken listing of prexies containing spaces?
Space becomes **%20** and **%20** becomes **%2520**

DEBUG o.apache.hadoop.fs.s3a.S3AFileSystem - Stripping trailing '/' from s3://my.bucket/my folder/
DEBUG o.apache.hadoop.fs.s3a.S3AFileSystem - listFiles(s3://my.bucket/my%20folder, true)
DEBUG o.apache.hadoop.fs.s3a.S3AFileSystem - Recursive list of all entries under my%20folder/
DEBUG o.apache.hadoop.fs.s3a.S3AFileSystem - LIST List my.bucket:/my%20folder/ delimiter=null keys=5000 requester pays=false
DEBUG com.amazonaws.auth.AWS4Signer - AWS4 Canonical Request: '"GET
/
fetch-owner=false&list-type=2&max-keys=5000&prefix=my%2520folder%2F

**Pull request merged**